core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -624,7 +625,24 @@ public PlanWithProperties visitSimpleTableExecuteNode(SimpleTableExecuteNode nod
         private PlanWithProperties visitTableWriter(PlanNode node, Optional<PartitioningScheme> partitioningScheme, PlanNode source, PreferredProperties preferredProperties, TableWriterNode.WriterTarget writerTarget)
```

Comment:
don't you need to add exchanges below `MergeProcessorNode`?

Reply:
I don't understand the question, but the produced plan seems to be correct. We have these stages in the plan:
```
(scan[target], scan[source]) -> RightJoin -> MergeProcessor -> MergeWriter -> TableCommit
```
Take a look at the `supporting-mege.rst` which might help explain the structure.

Reply:
How do you make sure `RightJoin` does not participate in CBO?

Reply:
Why would that matter?  If the CBO is working properly it should not change the output of the join operation, so if it has a better plan that should be ok.  right?